File: src/main/java/jb/files/serialization/ModelReader.java
package jb.files.serialization;

import de.bwaldvogel.liblinear.Linear;
import de.bwaldvogel.liblinear.Model;
import jb.config.Opts;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import static jb.config.Constants.modelsPath;

public class ModelReader {

    public List<Model> read(Opts opts) {
        File modelsDir = new File(modelsPath);
        List<Model> clfs = new ArrayList<>();
        for (File file : modelsDir.listFiles()) {
            if (isRightData(opts, file) && hasRightAmountOfBaseClfs(opts, file) && !inPermutation(Integer.valueOf(file.getName().split("_")[1]), opts.getPermutation())) {
                try {
                    clfs.add(Linear.loadModel(file));
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return clfs;
    }

    private boolean inPermutation(Integer numberOfSubset, int[] permutation) {
        return numberOfSubset.equals(permutation[0]) || numberOfSubset.equals(permutation[1]);
    }

    private boolean hasRightAmountOfBaseClfs(Opts opts, File file) {
        return Integer.valueOf(file.getName().split("_")[file.getName().split("_").length - 1]).equals(opts.getNumberOfBaseClassifiers() + 2);
    }

    private boolean isRightData(Opts opts, File file) {
        return file.getName().startsWith(opts.getFilename().substring(0, 3));
    }

}

File: src/main/java/jb/files/serialization/ModelReader.kt
package jb.files.serialization

import de.bwaldvogel.liblinear.Linear
import de.bwaldvogel.liblinear.Model
import jb.config.Opts

import java.io.File
import java.io.IOException
import java.util.ArrayList

import jb.config.Constants.modelsPath

class ModelReader {

    fun read(opts: Opts): List<Model> {
        val modelsDir = File(modelsPath)
        val clfs = ArrayList<Model>()
        for (file in modelsDir.listFiles()!!) {
            if (isRightData(opts, file) && hasRightAmountOfBaseClfs(opts, file) && !inPermutation(Integer.valueOf(file.name.split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1]), opts.permutation!!)) {
                try {
                    clfs.add(Linear.loadModel(file))
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }
        return clfs
    }

    private fun inPermutation(numberOfSubset: Int, permutation: IntArray): Boolean {
        return numberOfSubset == permutation[0] || numberOfSubset == permutation[1]
    }

    private fun hasRightAmountOfBaseClfs(opts: Opts, file: File): Boolean {
        return Integer.valueOf(file.name.split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[file.name.split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().size - 1]) == opts.numberOfBaseClassifiers + 2
    }

    private fun isRightData(opts: Opts, file: File): Boolean {
        return file.name.startsWith(opts.getFilename().substring(0, 3))
    }

}
